### PR TITLE
check timestamp

### PR DESCRIPTION
**Pull request by mattn**
Current implementation does not check timestamp. So it accept invalid timestamp.

### CODE REVIEW

**Comment by mattn** [comment]
I'll add test for this.

**Comment by mattn** [comment]
I added test. But latest go-nostr have breaking changes that type of ConnectionError is now string. So I disabled TestServerShutdownWebsocket.

**Comment by fiatjaf** [comment]
By invalid you mean a timestamp in the future?

I think Relayer shouldn't disallow these, they should be disallowed by the individual implementations that use relayer (like the `basic`) one.

There are some use cases for accepting events in the future. A relay could, for example, accept them, but not display until their time comes, thus making a seamless "scheduled post" service.

**Comment by pbennett** [comment]
Shouldn't events in the past be disallowed as well?  Keys could be compromised, attacker deletes old events, and posts new 'replacements' with strategically altered content.  It seems like 'posting' back-dated events should be considered invalid as well?

**Comment by fiatjaf** [comment]
Again, this should be part of individual implementations policies. Relayer itself is trying to be a generic framework for implementing relays on top of.

There are use cases for accepting old events, such as, for example, receiving an entire archive of posts from someone copying those from some other relay.

**Comment by mattn** [comment]
Considering web clients like iris and snort, future timestamp posts will always sit at the top of the screen. This can be an attack, I think.

**Comment by fiatjaf** [comment]
Iris and Snort should fix on their side. Also if a user is following someone else that messes up with timestamps like that that must be because the user is ok with that.

**Comment by mattn** [comment]
make sense. FYI, nostr-rs-relay rejects invalid timestamp.

https://github.com/scsibug/nostr-rs-relay/blob/8e4e2d824bd7defb3c17a959ebc17ba03778ebdf/src/event.rs#L319-L333

**Comment by fiatjaf** [comment]
Yes, I think it is fine for relays to reject, I just don't think it fits Relayer per se since Relayer is trying to be flexible. I would welcome a PR for the `relayer/basic` relay, for example (but since no one is using it and it's just an example I don't think you should waste your time on that).

**Comment by mattn** [comment]
Related issue: https://github.com/nostr-protocol/nips/issues/407

File: handler_test.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	if err == nil {
+		t.Fatal("should be an error")
+	}
+
+	ev.Content = "test"
+	ev.CreatedAt = time.Now()
+	ev.Kind = nostr.KindTextNote
+	ev.Sign(sk)
+	_, err = client.Publish(context.Background(), ev)
+	if err != nil {
```
Comment (isaqueveras) [inline handler_test.go:62]:
```suggestion
	if _, err = client.Publish(context.Background(), ev); err != nil {
```

File: handler_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	} else {
+		t.Fatal(err)
+	}
+
+	ev.Content = "test"
+	ev.CreatedAt = time.Now().Add(80 * time.Minute)
+	ev.Kind = nostr.KindTextNote
+	ev.Sign(sk)
+	_, err = client.Publish(context.Background(), ev)
+	if err == nil {
```
Comment (isaqueveras) [inline handler_test.go:53]:
```suggestion
	if _, err = client.Publish(context.Background(), ev); err == nil {
```

File: handler_test.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	var ev nostr.Event
+	var sk string
+	nsec := "nsec180cvv07tjdrrgpa0j7j7tmnyl2yr6yr7l8j4s3evf6u64th6gkwsgyumg0"
+	if _, s, err := nip19.Decode(nsec); err != nil {
+		t.Fatal(err)
+	} else {
+		sk = s.(string)
+	}
```
Comment (isaqueveras) [inline handler_test.go:38]:
```suggestion
	}

	sk = s.(string)
```

**Comment by mattn** [comment]
Related nip https://github.com/nostr-protocol/nips/blob/master/22.md

**Pull request closed without merging**